### PR TITLE
Support RankingComparator and break ranking ties in case of having different attack stats

### PR DESCRIPTION
**Pull request by Mygod**
Many bug fixes included. Commented out non-functional `CalculateAllRanks` and `CalculateTopRanks`.

Removed nonsensical test cases like testing level 0 stuff.

Fixes https://github.com/UnownHash/gohbem/issues/9. Mirrors https://github.com/Mygod/ohbem/pull/16.

### CODE REVIEW

File: pvp_core.go [hunk 1]
```diff
@@ -100,56 +86,110 @@ func calculateRanks(stats PokemonStats, cpCap int, lvCap float64) ([16][16][16]R
...
+	}
+	d = a.Cp - b.Cp
+	if d > 0 {
+		return 1
+	}
+	if d < 0 {
+		return -1
+	}
+	return 0
+}
```
Comment (lenisko) [inline pvp_core.go:165]: Change to CamelCase and add docstrings.

File: structs.go [hunk 2]
```diff
@@ -23,6 +24,15 @@ type League struct {
...
 }
 
+type PvPRankingStats struct {
+	Attack float64
+	Value  float64
+	Level  float64
+	Cp     int
+	Index  int
+}
+type RankingComparator func(a, b *PvPRankingStats) int
```
Comment (lenisko) [inline structs.go:34]: missing docs

File: pvp_core.go [hunk 3]
```diff
@@ -100,56 +86,110 @@ func calculateRanks(stats PokemonStats, cpCap int, lvCap float64) ([16][16][16]R
...
 			}
 		}
 	}
 
-	sort.Sort(rankingSortableIndexed(sortedRanks[:]))
+	// use Slice over SliceStable for performance since we have index available to us
+	sort.Slice(sortedRanks[:count], func(i, j int) bool {
+		d := comparator(&sortedRanks[i], &sortedRanks[j])
+		return d < 0 || d == 0 && sortedRanks[i].Index < sortedRanks[j].Index
+	})
```
Comment (lenisko) [inline pvp_core.go:188]: I've used `Sort` as `Slice` was noticeably slower in my benchmarks.

```
$ git checkout refined-rankings ; go build ; go test -benchtime=60s -bench=BenchmarkQueryPvPRank ; git checkout master ; go build ; go test -benchtime=60s -bench=BenchmarkQueryPvPRank
BenchmarkQueryPvPRank-24                   30672           2367168 ns/op # master
BenchmarkQueryPvPRank-24                   26839           2691408 ns/op # refined-rankings
```

`refined-rankings` is 12% slower than `master`, it's not much as we hit cache very fast, but experiment a bit with sorting, you might get better results.

**Review by lenisko — CHANGES_REQUESTED** [review]

File: pvp_core.go (hunk 3 above)
Comment (Mygod) [reply pvp_core.go:188]: If `Slice` was slower than `Sort`, then it is a golang issue, not a me issue?

Comment (Mygod) [reply pvp_core.go:188]: Also according to your code, isn't `master` in fact 12% slower than `refined-rankings`?

Comment (lenisko) [reply pvp_core.go:188]: Yes, could be. Give it a try and run a bench when you'll get a moment.

Comment (lenisko) [reply pvp_core.go:188]: > Also according to your code, isn't `master` in fact 12% slower than `refined-rankings`?

[Nope.](https://www.practical-go-lessons.com/chap-34-benchmarks#how-to-read-benchmark-results)

Comment (Mygod) [reply pvp_core.go:188]: https://github.com/golang/go/issues/22504#issuecomment-340685108

Golang sucks so bad. How is "improving documentations" not actionable?

**Comment by Mygod** [comment]
Before changing to `sort.Sort`:
```
BenchmarkQueryPvPRank-16                   22726           3200570 ns/op
BenchmarkQueryPvPRankCached-16          64416926              1072 ns/op
```

After:
```
BenchmarkQueryPvPRank-16                   26271           2741242 ns/op
BenchmarkQueryPvPRankCached-16          61786150              1050 ns/op
```

Master branch:
```
BenchmarkQueryPvPRank-16                   24817           2860125 ns/op
BenchmarkQueryPvPRankCached-16          43727300              1685 ns/op
```

File: pvp_core.go (hunk 3 above)
Comment (Mygod) [reply pvp_core.go]: Also created a PR at golang. https://github.com/golang/go/pull/59449

**Review by lenisko — APPROVED** [review]

**Pull request merged**